pkg/compact/compact.go

```diff
@@ -878,28 +880,66 @@ func (cg *Group) deleteBlock(b string) error {
 
 // BucketCompactor compacts blocks in a bucket.
 type BucketCompactor struct {
-	logger     log.Logger
-	sy         *Syncer
-	comp       tsdb.Compactor
-	compactDir string
-	bkt        objstore.Bucket
+	logger      log.Logger
+	sy          *Syncer
+	comp        tsdb.Compactor
+	compactDir  string
+	bkt         objstore.Bucket
+	concurrency int
 }
 
 // NewBucketCompactor creates a new bucket compactor.
-func NewBucketCompactor(logger log.Logger, sy *Syncer, comp tsdb.Compactor, compactDir string, bkt objstore.Bucket) *BucketCompactor {
+func NewBucketCompactor(logger log.Logger, sy *Syncer, comp tsdb.Compactor, compactDir string, bkt objstore.Bucket, concurrency int) *BucketCompactor {
 	return &BucketCompactor{
-		logger:     logger,
-		sy:         sy,
-		comp:       comp,
-		compactDir: compactDir,
-		bkt:        bkt,
+		logger:      logger,
+		sy:          sy,
+		comp:        comp,
+		compactDir:  compactDir,
+		bkt:         bkt,
+		concurrency: concurrency,
 	}
 }
 
 // Compact runs compaction over bucket.
 func (c *BucketCompactor) Compact(ctx context.Context) error {
 	// Loop over bucket and compact until there's no work left.
 	for {
+		// Set up workers who will compact the groups when the groups are ready.
+		var (
+			errGroup, errGroupCtx = errgroup.WithContext(ctx)
+			groupChan             = make(chan *Group)
+			finishedAllGroups     = true
+			mtx                   sync.Mutex
+		)
+
+		for i := 0; i < c.concurrency; i++ {
+			errGroup.Go(func() error {
+				for g := range groupChan {
+					level.Info(c.logger).Log("msg", "worker doing some work")
```

Comment:
That's kind of vague log message. We can make it debug and write something meaningful (: I think it's debug leftover

Reply:
it was a debug leftover, sorry :)  removed now